Sensor

CSTaylorJr opened the pull request:
Added the low level sensor stuff. Still need to convert raw flow sensor data to real world units. Code works, uses serial debug window and on-board LED as debug outputs.

CSTaylorJr committed:
Initial checking of sensor HAL. Have bug somewhere in threading code.

CSTaylorJr committed:
Changed to sized data types (int16_t instead if int), moved pressure calc to HAL, fixed bug in serial debug output.

mrrosen commented on src/hal/hal.h:
```diff
@@ -2,6 +2,8 @@
 #ifndef __HAL_H__
 #define __HAL_H__
 
+#include <arduino.h>
```
Dont modify this file please

CSTaylorJr replied:
Moved to hal/sensor.cpp

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -60,6 +55,10 @@ PT_THREAD(sensorsThreadMain(struct pt* pt))
   PT_BEGIN(pt);
 
+  PT_WAIT_UNTIL(pt, (millis()-last_debug_time)>250);
+
```
Can we get rid of the debug code in the pr?

CSTaylorJr replied:
Sure. Not certain what you want to do instead.

CSTaylorJr replied:
Done!

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -20,28 +19,24 @@ static struct pt sensorsPressureThread;
 {
   PT_BEGIN(pt);
-
-  PT_WAIT_UNTIL(pt, pressureSensorHalFetch() != HAL_IN_PROGRESS);
```
In order to make sure this doesnt spam too hard, can we add a `timerHalBegin(&pressureTimer, PRESSURE_SAMPLING_RATE);` then a `PT_WAIT_UNTIL(pt, timerHalRun() != HAL_IN_PROGRESS)` run before the thread resets

CSTaylorJr replied:
Done

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -20,28 +19,24 @@ static struct pt sensorsPressureThread;
   PT_BEGIN(pt);
-  
-  // TODO: airflow sensor
-
```
same thing about timer stuff here

CSTaylorJr replied:
Done

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -72,23 +71,35 @@ PT_THREAD(sensorsThreadMain(struct pt* pt))
+  Serial.print("    ");
+  Serial.print(sensors.currentFlow);
+  Serial.println();
+
```
remove debug code

CSTaylorJr replied:
Same thing--just let me know how you want it done! :)

CSTaylorJr replied:
Done!

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -72,23 +71,35 @@ PT_THREAD(sensorsThreadMain(struct pt* pt))
+  pinMode(LED_BUILTIN, OUTPUT);
+  // Serial debug port
+  Serial.begin(38400);
+  Serial.print("Start");
```
remove debug code

mrrosen commented on src/hal/sensor/sensor.cpp:
```diff
@@ -0,0 +1,80 @@
+  TCNT5=0;  // initialize counter to 0
+  OCR5A= 1000;  // Set interrupt to 2khz
+  TIMSK5 |= 0x02; // enable timer compare interrupt
+
```
Since we want to sample the sensors at 40Hz each (is that fast enough to oversample?), why not just run the timer at 80Hz (change prescaler at /64, set OCR5A to 3125, changing `timer5Counter` to a bit flip instead)? saves many unneeded interrupts.

mrrosen replied:
Also; can we use the same style used in the motor control for setting bits, I know its verbose but its somewhat easier to debug without having to refer to the datasheet for the bit position of each setting)?

MyElectrons replied:
The flow sensor results get integrated into the volume pseudo-sensor. Initial tests showed good repeatability at 100Hz or 10ms sampling intervals. The lower the flow sampling frequency is - the greater error/noise we'll be getting as volume measurements. Unless we are cycles constrained (where?) I'd suggest we keep reading flow and integrating the volume at 100Hz.

mrrosen replied:
I think I miscalculated the timer speed sorry; it is set to 2KHz and the timer isr further divides that by 20. I was just suggesting to make the timer run at 200Hz directly and divide the timer isr by 2 instead, saving 18 unneeded isrs. I somehow got to a sample rate of 40Hz; and not Im not sure what math got me there. So, make `OCR5A = 10000` and make the counter a boolean flag.

mrrosen commented on src/hal/sensor/sensor.cpp:
```diff
@@ -0,0 +1,80 @@
+// This routine returns the pressure value out of the filter
+int16_t pressureSensorHalGetValue(int16_t *value){
+  int32_t temp = pSum>>4;
+  int32_t pascals = ((temp * 217L) - 110995L)>>2;	// convert to Pascals
```
Need to wrap the grabbing of pSum in an `ATOMIC_BLOCK` to avoid problems of interrupts coming in between the bytes since this is an 8bit processor after all.
```
  // Standard avr-libc include for atomic operations
  #include <util/atomic.h>
  ...
  ATOMIC_BLOCK(ATMOIC_RESTORESTATE) {
  int32_t temp = pSum>>4;
  }
  int32_t pascals = ((temp * 217L) - 110995L)>>2;
```

CSTaylorJr replied:
Done

mrrosen commented on src/hal/sensor/sensor.cpp:
```diff
@@ -0,0 +1,80 @@
+}
+// This routine returns the flow sensor value out of the filter
+int16_t airflowSensorHalGetValue(int16_t *value){
+  *value = fSum>>3;
```
See above about atomic access ,`fSum` needs it too

CSTaylorJr replied:
Done

mrrosen commented on src/hal/sensor/sensor.cpp:
```diff
@@ -0,0 +1,80 @@
+//    running through a /8 filter
+//  For now, I have the serial print triggering every second.
+int16_t timer5Count=0;
+ISR(TIMER5_COMPA_vect){
```
Some comments above about reducing timer isr, changes this stuff too

CSTaylorJr replied:
Do you mean that it doesn't need to run at a 0.5ms cadence. That's certainly true. I can slow it down to 5ms for now, and we can speed it up later if we get sensors that have to be collected faster?

mrrosen replied:
Yes, thats what I meant, see above (I mismath'd)

CSTaylorJr replied:
Added.

mrrosen commented on src/hal/sensor/sensor.h:
```diff
@@ -0,0 +1,17 @@
+int16_t sensorHalInit(void);
+
+// Gets the airflow sensor value
+int16_t airflowSensorHalGetValue(int16_t *value);
```
Return codes can be int; values need definite size, return codes are ok as `int` (unless I later want to make them their own type....)

CSTaylorJr replied:
Done

mrrosen commented on src/hal/sensor/sensor.h:
```diff
@@ -0,0 +1,17 @@
+
+#ifndef __SENSOR_HAL_H__
```
Im not sure if I dont want to keep two separate header files and we can just have a single .cpp to implement them; since its not clear if wed want to combine them in all situations

CSTaylorJr replied:
I'm okay either way. Just let me know.

CSTaylorJr replied:
After discussion, leaving it as is for now.

mrrosen reviewed (changes requested):
Initial review; several suggestions and thoughts

CSTaylorJr committed:
Changes based on Mike's feedback.

MyElectrons reviewed (commented):
Please try to keep reading flow and integrating volume at 100Hz.

mrrosen commented on src/hal/sensor/sensor.cpp:
```diff
@@ -1,14 +1,16 @@
 
 #include "../../hal/hal.h"
 #include "../../hal/sensor/sensor.h"
+#include <arduino.h>
```
Im just noticing this now, but it should be `Arduino.h` with a capital A. Google suggests that Windows isnt case sensitive in include file names but Linux is, so we should try to avoid that
```suggestion
#include <Arduino.h>
```

CSTaylorJr replied:
Fixed!

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -18,22 +20,30 @@ static struct pt sensorsThread;
 
 uint32_t last_debug_time;
 
+#define PRESSURE_SAMPLING_RATE 50000
```
Can we move this `#define`s to the top of the file. Also, please use the `MSEC` units so its more clear the units:
```suggestion
#define PRESSURE_SAMPLING_RATE (50 MSEC)
```

CSTaylorJr replied:
When I did that, I got an overflow warning.

mrrosen replied:
Really? these should be the same since 50 MSEC should be preprocessed to 50 * 1000

CSTaylorJr replied:
I know--it's very strange. I figured I'd leave it 'til we got the timer stuff updated.

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -18,22 +20,30 @@ static struct pt sensorsThread;
 {
   PT_BEGIN(pt);
+  timerHalBegin(&pressureTimer, PRESSURE_SAMPLING_RATE);  // set to 50ms to not overflow timer. Can be slowed down once timer hal is updated.
+  PT_WAIT_UNTIL(pt, timerHalRun(&pressureTimer)!=HAL_IN_PROGRESS);
```
If we put this right before `PT_RESTART` any time taken to get the sample and do the processing will be included in the timeout, making the sample rate more consistent (though barely, I dont expect it to take more than 1ms to do everything, but maybe it might take 2ms in the worst case if our math for derived information gets fancy)

CSTaylorJr replied:
Did you mean to move the PT_RESTART up to right beneath the PT_WAIT_UNTIL? I tried that, and it mucked things up with the debug reporting. Maybe we can sync and get the debug stuff working properly, and then this should work fine?

CSTaylorJr replied:
Fixed after Mike set me straight. Done.

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -18,22 +20,30 @@ static struct pt sensorsThread;
 {
   PT_BEGIN(pt);
+  timerHalBegin(&airflowTimer, AIRFLOW_SAMPLING_RATE);  // set to 50ms to not overflow timer. Can be slowed down once timer hal is updated.
+  PT_WAIT_UNTIL(pt, timerHalRun(&airflowTimer)!=HAL_IN_PROGRESS);
```
See above

mrrosen replied:
See above

mrrosen commented on src/modules/sensors.h:
```diff
@@ -32,9 +32,9 @@ struct sensors {
 
 // TODO: Doc
-int sensorsModuleRun(void);
+int16_t sensorsModuleRun(void);
```
Please make these `int` instead of `int16_t`

CSTaylorJr replied:
Done

mrrosen reviewed (changes requested):
A few minor changes

mrrosen commented:
I did another review and tried to respond so some of the comments made by Calvin and Sergey, please review and fix where needed

CSTaylorJr committed:
Flow & Volume added

Added calculations for real world flow and volume measurements

CSTaylorJr committed:
Fixes based on review

Made fixes based on review comments. The big change is that timer5 is slowed to a cadence of 5ms.

CSTaylorJr committed:
Minor changes based on Mike's feedback

CSTaylorJr committed:
Fixed debug messaging for sensors.

Commit:
sensor: refactor flow/volume sensor calculations

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
sensors: modify sampling times to use timer units

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
sensor: fix some math problems in volume calculation

Specifically, typo in sampling rate led to erroneous maximum error
calculation; ultimately increased precision of the accumulator and
had to increase its size to an int32_t

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
sensor: fix unit math (and comment) for airvolume

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
sensors: fix up a few things in the sensor driver and module

. Remove filter from sensor readings, will need to reevaluate
  once we have more development on closed loop and better
  understand the system noise
. Make volume integration reset API 'public'
. Fix up some printing and sampling rate from the sensors for
  initial merge

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>
